File: capa1/simuladores/python/src/mqtt_publisher.py
import json
import logging
import time
import paho.mqtt.client as mqtt

logger = logging.getLogger(__name__)


class MQTTPublisher:
    def __init__(self, broker_host: str, broker_port: int = 1883,
                 client_id: str = "urbia-sim"):
        self.broker_host = broker_host
        self.broker_port = broker_port
        self._conectado  = False
        self._publicados = 0
        self._errores    = 0

        self.client = mqtt.Client(
            mqtt.CallbackAPIVersion.VERSION2,
            client_id=client_id,
            clean_session=False,
        )
        self.client.on_connect    = self._on_connect
        self.client.on_disconnect = self._on_disconnect
        self.client.on_publish    = self._on_publish

    def conectar(self):
        self.client.connect_async(self.broker_host, self.broker_port, keepalive=60)
        self.client.loop_start()
        time.sleep(1)

    def publicar(self, topic: str, payload: dict) -> bool:
        if not self._conectado:
            self._errores += 1
            return False
        try:
            msg = json.dumps(payload, ensure_ascii=False)
            result = self.client.publish(topic, msg, qos=1)
            if result.rc == mqtt.MQTT_ERR_SUCCESS:
                return True
            self._errores += 1
            return False
        except Exception as e:
            self._errores += 1
            logger.error(f"Error publicando: {e}")
            return False

    def desconectar(self):
        self.client.loop_stop()
        self.client.disconnect()

    def _on_connect(self, client, userdata, flags, reason_code, properties):
        self._conectado = not reason_code.is_failure
        if self._conectado:
            logger.info(f"Conectado a {self.broker_host}:{self.broker_port}")
        else:
            logger.error(f"Error de conexión: {reason_code}")

    def _on_disconnect(self, client, userdata, flags, reason_code, properties):
        self._conectado = False
        if reason_code.value != 0:
            logger.warning(f"Desconexión inesperada ({reason_code}), reconectando...")

    def _on_publish(self, client, userdata, mid, reason_code, properties):
        self._publicados += 1

    def estadisticas(self) -> dict:
        return {
            "conectado":  self._conectado,
            "publicados": self._publicados,
            "errores":    self._errores,
        }

File: capa1/simuladores/cpp/src/mqtt_publisher.cpp
#include "mqtt_publisher.hpp"
#include <iostream>
#include <stdexcept>

MQTTPublisher::MQTTPublisher(const std::string& broker_host, int broker_port,
                             const std::string& client_id)
    : client_id(client_id) {
    broker_uri = "tcp://" + broker_host + ":" + std::to_string(broker_port);
    client = std::make_unique<mqtt::async_client>(broker_uri, client_id);
}

MQTTPublisher::~MQTTPublisher() {
    if (_conectado) {
        try { desconectar(); } catch (...) {}
    }
}

void MQTTPublisher::conectar() {
    auto opts = mqtt::connect_options_builder()
        .clean_session(false)
        .keep_alive_interval(std::chrono::seconds(60))
        .automatic_reconnect(std::chrono::seconds(2), std::chrono::seconds(30))
        .finalize();

    try {
        client->connect(opts)->wait();
        _conectado = true;
        std::cout << "[MQTT] Conectado a " << broker_uri << "\n";
    } catch (const mqtt::exception& e) {
        throw std::runtime_error(std::string("Error MQTT conectar: ") + e.what());
    }
}

bool MQTTPublisher::publicar(const std::string& topic, const nlohmann::json& payload) {
    if (!_conectado) {
        ++_errores;
        std::cerr << "[MQTT] Sin conexión — mensaje descartado\n";
        return false;
    }
    try {
        auto msg = mqtt::make_message(topic, payload.dump(), 1, false);
        client->publish(msg)->wait_for(std::chrono::seconds(5));
        ++_publicados;
        return true;
    } catch (const mqtt::exception& e) {
        ++_errores;
        std::cerr << "[MQTT] Error publicando en " << topic << ": " << e.what() << "\n";
        return false;
    }
}

void MQTTPublisher::desconectar() {
    try {
        client->disconnect()->wait();
        _conectado = false;
        std::cout << "[MQTT] Desconectado limpiamente\n";
    } catch (const mqtt::exception& e) {
        std::cerr << "[MQTT] Error al desconectar: " << e.what() << "\n";
    }
}

MQTTPublisher::Stats MQTTPublisher::estadisticas() const {
    return {_conectado ? 1 : 0, _publicados.load(), _errores.load()};
}
